### PR TITLE
[libc++] [test] Fix portability issues for MSVC

### PR DESCRIPTION
**Pull request by StephanTLavavej**
* Guard `std::__make_from_tuple_impl` tests with `#ifdef _LIBCPP_VERSION` and `LIBCPP_STATIC_ASSERT`.
* Change `_LIBCPP_CONSTEXPR_SINCE_CXX20` to `TEST_CONSTEXPR_CXX20`.
  + Other functions in `variant.swap/swap.pass.cpp` were already using the proper test macro.
* Mark `what` as `[[maybe_unused]]` when used by `TEST_LIBCPP_REQUIRE`.
  + This updates one occurrence in `libcxx/test/libcxx` for consistency.
* Windows `_putenv_s()` takes 2 arguments, not 3.
  + See [MSVC documentation][].
  + POSIX `setenv()` takes `int overwrite`, but Windows `_putenv_s()` always overwrites.
* Avoid non-Standard zero-length arrays.
  + Followup to #74183 and #79792.
* Add `operator++()` to `unsized_it`.
  + The Standard requires this due to [N4981][] \[move.iter.requirements\]/1 "The template parameter `Iterator` shall either meet the *Cpp17InputIterator* requirements (\[input.iterators\]) or model `input_iterator` (\[iterator.concept.input\])."
  + MSVC's STL requires this because it has a strengthened exception specification in `move_iterator` that inspects the underlying iterator's increment operator.
* `uniform_int_distribution` forbids `int8_t`/`uint8_t`.
  + See [N4981][] \[rand.req.genl\]/1.5. MSVC's STL enforces this.
  + Note that when changing the distribution's `IntType`, we need to be careful to preserve the original value range of `[0, max_input]`.
* fstreams are constructible from `const fs::path::value_type*` on wide systems.
  + See \[ifstream.cons\], \[ofstream.cons\], \[fstream.cons\].
* In `msvc_stdlib_force_include.h`, map `_HAS_CXX23` to `TEST_STD_VER` 23 instead of 99.
  + On 2023-05-23, https://github.com/llvm/llvm-project/commit/71400505ca048507e827013eb1ea0bc863525cab started recognizing 23 as a distinct value.
* Fix test name typo: `destory_elements.pass.cpp` => `destroy_elements.pass.cpp`

[N4981]: https://wg21.link/N4981
[MSVC documentation]: https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/putenv-s-wputenv-s?view=msvc-170

### CODE REVIEW

**Review by ldionne — APPROVED** [review]

File: libcxx/test/std/numerics/numeric.ops/numeric.ops.gcd/gcd.pass.cpp [hunk 1]
```diff
@@ -67,12 +68,14 @@ T basic_gcd(T m, T n) {
 template <typename Input>
 void do_fuzzy_tests() {
   std::mt19937 gen(1938);
-  std::uniform_int_distribution<Input> distrib;
+  using DistIntType         = std::conditional_t<sizeof(Input) == 1, int, Input>; // See N4981 [rand.req.genl]/1.5
+  constexpr Input max_input = std::numeric_limits<Input>::max();
+  std::uniform_int_distribution<DistIntType> distrib(0, max_input);
```
Comment (cjdb) [inline libcxx/test/std/numerics/numeric.ops/numeric.ops.gcd/gcd.pass.cpp:73]: We need to keep `Input` as is. Your commit message says that the standard prohibits `int8_t`, but the wording is "undefined", not "ill-formed", meaning that we're allowed to provide `int8_t` as a documented extension, which libc++ does.

Comment (ldionne) [reply libcxx/test/std/numerics/numeric.ops/numeric.ops.gcd/gcd.pass.cpp:73]: Sorry @cjdb , I merged before you had time to comment. This wasn't intentional.

Comment (StephanTLavavej) [reply libcxx/test/std/numerics/numeric.ops/numeric.ops.gcd/gcd.pass.cpp:73]: But this is the `libcxx/test/std` subdirectory, which is portable. If you want to test a libc++ extension, that should be guarded with the libc++ macro (which I would be fine with).

Comment (cjdb) [reply libcxx/test/std/numerics/numeric.ops/numeric.ops.gcd/gcd.pass.cpp:73]: Oh, this isn't testing `uniform_int_distribution`? Never mind then.

**Pull request merged**